pkg/callerinfo/callerinfo.go

```diff
@@ -0,0 +1,108 @@
+// Copyright 2023 Outreach Corporation. All Rights Reserved.
+
+// Description: Provides the ability to discover the module+function who is calling you.
+
+// Package callerinfo provides the GetCallerFunction to get the name of the module and function
+// that has called you.
+package callerinfo
+
+import (
+	"fmt"
+	"runtime"
+	"runtime/debug"
+	"strings"
+	"sync"
+)
+
+type CallerInfo struct {
+	Function      string
+	File          string
+	LineNum       uint
+	Module        string
+	ModuleVersion string
+}
+
+var moduleLookupByPC = make(map[uintptr]CallerInfo)
+var moduleLookupLock = sync.RWMutex{}
+
+var buildInfo *debug.BuildInfo
+
+//nolint:gochecknoinits // Why: Because we need it and the side effects are self-contained and non-mutating
+func init() {
+	if bi, ok := debug.ReadBuildInfo(); ok {
+		buildInfo = bi
+	}
+}
+
+// GetCallerInfo returns various types of info about the caller of the function.
+// skipFrames determines how many frames above the caller to skip.  If you want to know who is calling your
+// function, you would pass in skipFrames of 1.  skipFrames of 0 will return yourself.
+func GetCallerInfo(skipFrames uint16) (CallerInfo, error) {
+	// We only care about the one stack frame above the caller, so skip at least two:
+	// 1. runtime.Callers
+	// 2. GetCallerModule
+	pc := make([]uintptr, 1)
+	skipTotal := 2 + int(skipFrames)
+	num := runtime.Callers(skipTotal, pc)
+	if num == 0 {
+		return CallerInfo{}, fmt.Errorf("no frames returned from skip %d", skipTotal)
+	}
+
+	if mod, valid := moduleLookupByPC[pc[0]]; valid {
+		// Found it in the cache
+		return mod, nil
+	}
+
+	// Not cached -- have to do the slow lookup
+	frames := runtime.CallersFrames(pc)
+	frame, _ := frames.Next()
+
+	ci := CallerInfo{
+		Function: frame.Function,
+		File:     frame.File,
+		LineNum:  uint(frame.Line),
+	}
+
+	// Attempt to back-calc module name by searching through deps
+	if buildInfo != nil {
```

Comment:
def move this into another function, having it be nested makes it harder to read here to begin w/ but also it fits into it's own function p easily

Reply:
okie done